### PR TITLE
Set knowledge base slots

### PR DESCRIPTION
**Pull request by tabergma**
Related PR https://github.com/RasaHQ/rasa-sdk/pull/96

**Proposed changes**:
- automatically set internal slots for knowledge base action when action name is present (prefix `knowledge_base`)
- add example bot that interacts with a knowledge base
- add documentation page for knowledge base action

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -16,18 +16,21 @@ Changed
...
   information.
-- The tracker is now available within the interpreter's ``parse`` method, giving the ability to create interpreter classes that 
-  use the tracker state (eg. slot values) during the parsing of the message. More details on motivation of this change see issues/3015
-
-Changed
--------
+- The tracker is now available within the interpreter's ``parse`` method, giving the ability to create interpreter
+  classes that use the tracker state (eg. slot values) during the parsing of the message. More details on motivation
+  of this change see issues/3015
+- The tracker is now available within the interpreter's ``parse`` method, giving the ability to create interpreter
```
Comment (amn41) [inline CHANGELOG.rst]: duplicate of above

File: docs/core/knowledge-bases.rst [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+.. _knowledge_bases:
+
+Knowledge Base Actions
+======================
+
+.. edit-link::
+
+.. warning::
+   This feature is experimental.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
   This feature is experimental. We introduce experimental features to get feedback on implementations, so do please try it out. However, the functionality might be changed or removed in the future. If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_
```

File: docs/core/knowledge-bases.rst [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+.. contents::
+   :local:
+
+A lot of users want to obtain detailed information about certain objects, such as restaurants or hotels, during a conversation.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: I think we can make this easier to understand by motivating with an example. Otherwise this sounds very abstract

File: docs/core/knowledge-bases.rst [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+The action is also able to resolve certain mention of objects, such as ``the first one`` or ``that restaurant``.
+You can find a complete example in ``examples/knowledge_base_bot``.
+
+.. note::
+   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
+
+
+Create a KnowledgeBase
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
Create a Knowledge Base
```

File: docs/core/knowledge-bases.rst [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                "swimming-pool": false
+            },
+        ]
+    }
+
+The above json file contains data about restaurants and hotels.
+The json structure should contain a key for every object type, i.e. "restaurant" and "hotel".
+Every object type maps to a list of objects.
+
+Once the data are defined in a json file, called, for example, ``data.json``, you can create your
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: did we get rid of the `schema` part? if so that's a nice simplification

File: docs/core/knowledge-bases.rst [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+Customize your InMemoryKnowledgeBase
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The class ``InMemoryKnowledgeBase`` inherits ``KnowledgeBase``.
+You can customize your ``InMemoryKnowledgeBase`` by overwriting the following functions:
+
+- ``get_key_attribute_of_object``: To keep track of what object the user was talking about last, we store the value
+  of the key attribute in a specific slot. Every object should have a key attribute that is unique, i.e.
+  similar to the primary key in a relation database. Per default the name of the key attribute for every object type
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
  similar to the primary key in a relation database. By default the name of the key attribute for every object type
```

File: docs/core/knowledge-bases.rst [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+  is set to "id". You can overwrite the name of the key attribute for a specific object type by calling
+  ``set_key_attribute_of_object()``.
+- ``get_representation_function_of_object``: This methods returns a lambda function that maps an object from the
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: this is a difficult sentence (part of a difficult paragraph). Let's split it up into smaller sentences and give quick examples

File: docs/core/knowledge-bases.rst [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+  knowledge base to a string representation. This function is used whenever an object is outputted to the user.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: 'outputted' sounds awkward. what are we actually trying to say?

File: docs/core/knowledge-bases.rst [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+  Per default the lambda function is set to ``lambda obj: obj["name"]``. So, it returns the value of the attribute
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
  By default the lambda function is set to ``lambda obj: obj["name"]``. So, it returns the value of the attribute
```

changed to "By default" - "Per" is also correct but less common

File: docs/core/knowledge-bases.rst [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+Every object type maps to a list of objects.
+
+Once the data are defined in a json file, called, for example, ``data.json``, you can create your
+``InMemoryKnowledgeBase``:
+
+.. code-block:: python
+
+    knowledge_base = InMemoryKnowledgeBase.load("data.json")
+
+Customize your InMemoryKnowledgeBase
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: I feel like this section contains a lot of advanced details that most people don't need, correct? If I were looking to get started with this feature I would find all of this intimidating

File: docs/core/knowledge-bases.rst [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+          "7": lambda l: l[6],
+          "8": lambda l: l[7],
+          "9": lambda l: l[8],
+          "10": lambda l: l[9],
+          "ANY": lambda l: random.choice(list),
+          "LAST": lambda l: l[-1],
+      }
+
+  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
+  More on the usage can be found in section :ref:`resolve_mentions`.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: Avoid passive voice like "More on the usage can be found". Instead say:
"If you want to learn more about X, go to ..."

File: docs/core/knowledge-bases.rst [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+Just inherit ``KnowledgeBase`` and implement the methods ``get_objects()``, ``get_object()``, and
+``get_attributes_of_object()``.
+You can also customize your knowledge base further, for example, by adapting the methods mentioned in the previous
+section.
+
+.. note::
+   We wrote a `blog post <https://blog.rasa.com/set-up-a-knowledge-base-to-encode-domain-knowledge-for-rasa/>`_
+   that explains how you can set up your own knowledge base.
+
+Create an ActionQueryKnowledgeBase
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: I'd move this further down, below the NLU data description. If people want to set up a "proper" KB they can skip to that section

File: docs/core/knowledge-bases.rst [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+Or if no entities could be found
+
+    `I could not find any objects of type 'restaurant'.`
+
+If you want to change the utterance of the bot, you can overwrite the method ``utter_objects()`` in your action.
+
+Query the Knowledge Base for an Attribute of an Object
+------------------------------------------------------
+
+To obtain the value of an attribute for a specific object from the knowledge base, the action needs to know the object
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: This is super difficult language: "To obtain the value of an attribute for a specific object" - I think we can say this in plain english by usign an example

File: docs/core/knowledge-bases.rst (hunk 5 above)
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: yes, split it up into separate methods and set default values

File: docs/core/knowledge-bases.rst (hunk 10 above)
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: depends. if the user is using the default values, e.g. every object has an id and a name field, he does not need to do anything. if not then he should overwrite those methods.

File: examples/knowledgebasebot/actions.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class ActionMyKB(ActionQueryKnowledgeBase):
+    def __init__(self):
+        # load knowledge base with data from the given file
+        knowledge_base = InMemoryKnowledgeBase.load("knowledge_base_data.json")
+
+        # overwrite the representation function of the hotel object
+        # by default the representation function is just the name of the object
+        knowledge_base.set_representation_function_of_object(
```
Comment (tmbo) [inline examples/knowledgebasebot/actions.py]: is `object` the right word? maybe rather`type` or something from the knowledge base domain - e.g. how does grakn call them?

File: examples/knowledgebasebot/data/nlu.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+## intent:goodbye
+- bye
+- goodbye
+- see you around
+- see you later
+
+## intent:query_knowledge_base
+- what [restaurants](object_type:restaurant) can you recommend?
+- list some [restaurants](object_type:restaurant)
+- can you name some [restaurants](object_type:restaurant) please?
```
Comment (tmbo) [inline examples/knowledgebasebot/data/nlu.md]: is `object_type` necessary or would `type` suffice?

File: examples/knowledgebasebot/domain.yml [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  - text: "Hey!"
+  - text: "Hello! How can I help you?"
+
+  utter_goodbye:
+  - text: "Bye"
+  - text: "Goodbye. See you soon."
+
+  utter_ask_rephrase:
+  - text: "Sorry, I'm not sure I understand. Can you rephrase?"
+  - text: "Can you please rephrase? I did not got that."
```
Comment (tmbo) [inline examples/knowledgebasebot/domain.yml]:
```suggestion
  - text: "Can you please rephrase? I did not got that."

```

File: examples/knowledgebasebot/endpoints.yml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+action_endpoint:
+  url: "http://localhost:5055/webhook"
```
Comment (tmbo) [inline examples/knowledgebasebot/endpoints.yml]:
```suggestion
  url: "http://localhost:5055/webhook"

```

File: examples/knowledgebasebot/knowledge_base_data.json [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+            "name": "Berlin Hotel",
+            "price-range": "mid-range",
+            "breakfast-included": true,
+            "city": "Berlin",
+            "free-wifi": true,
+            "star-rating": 3,
+            "swimming-pool": true
+        }
+    ]
+}
```
Comment (tmbo) [inline examples/knowledgebasebot/knowledge_base_data.json]:
```suggestion
}

```

File: docs/core/knowledge-bases.rst [hunk 19]
```diff
@@ -0,0 +1,416 @@
...
+Knowledge base actions enable you to handle the following kind of dialogues:
+
+.. image:: ../_static/images/knowledge-base-example.png
+
+As you can see users might want to obtain detailed information about certain objects, such as restaurants or hotels,
+during a conversation.
+They want to obtain details, such as, if an restaurant has outside seating or how expensive the restaurant is.
+In order to answer those user requests domain knowledge is needed.
+Hard-coding the information would not help as the information are subject to change.
+Additionally, users do not only refer to objects by their names, but also use terms, such as "the first one" or "it",
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: I think this is the key point that the KB helps with so should be mentioned first.

File: docs/core/knowledge-bases.rst [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+to refer to a specific restaurant.
+Restaurants need to be recognised and reused at a later point in the conversation.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: I'm not sure what this sentence is telling me

File: examples/knowledgebasebot/actions.py (hunk 14 above)
Comment (tabergma) [reply examples/knowledgebasebot/actions.py]: grakn calls them entities, but entities is confusing as we already use the term entities for our NER. Alternative terms would be nodes, but nodes is only related to graph databases and not used in relation databases. So I went with object. Let me know if you have a better idea.

File: docs/core/knowledge-bases.rst [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+
+To handle the above challenges, we recommend that you create a custom action that inherits from ``ActionQueryKnowledgeBase``.
+This is a single actions which contains the logic to query a knowledge base for objects and their attributes.
+The action is also able to resolve certain mention of objects, such as ``the first one`` or ``that restaurant``.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
When a restaurant is mentioned indirectly, for example using a phrase like ``the first one`` or ``that restaurant``, this action is able to figure out which restaurant the user is referring to.
```

File: docs/core/knowledge-bases.rst (hunk 8 above)
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: bot is referring to an object. rephrased it.

File: docs/core/knowledge-bases.rst [hunk 22]
```diff
@@ -0,0 +1,416 @@
...
+  object type listed in the knowledge base, e.g. "restaurant".
+- ``mention``: If the user refers to an object via "the first one", "that one", or "it", you should mark those terms
+  as ``mention``. We also use :ref:`entity_synonyms` to map some of the mentions to symbols. More on that in
+  :ref:`resolve_mentions`.
+- ``attribute``: All attribute names defined in your knowledge base should be marked as ``attribute`` in the NLU data.
+  Again, use :ref:`entity_synonyms` to map variations of an attribute name to the one used in the knowledge base.
+
+Don't forget to add those entities to your domain file (as entities and slots).
+
+
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion

.. _custom_knowledge_base:


```

File: docs/core/knowledge-bases.rst [hunk 23]
```diff
@@ -0,0 +1,416 @@
...
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
+
+
+Create a Knowledge Base
+-----------------------
+
+The data you will use to answer the user's request comes from a knowledge base.
+A knowledge base can be used to store complex data structures.
+If you just have some data points that fit into memory, you can use our ``InMemoryKnowledgeBase`` implementation.
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
We suggest you get started by using the ``InMemoryKnowledgeBase``. Once you want to start working with a large amount of data, you can switch to a :ref:`custom-knowledge-base` .
```

File: examples/knowledgebasebot/actions.py [hunk 24]
```diff
@@ -0,0 +1,16 @@
+from rasa_sdk.knowledge_base.storage import InMemoryKnowledgeBase
+from rasa_sdk.knowledge_base.actions import ActionQueryKnowledgeBase
+
+
+class ActionMyKB(ActionQueryKnowledgeBase):
+    def __init__(self):
+        # load knowledge base with data from the given file
+        knowledge_base = InMemoryKnowledgeBase.load("knowledge_base_data.json")
```
[the remark this answers is not in the document]
Comment (tabergma) [reply examples/knowledgebasebot/actions.py]: So 
```
knowledge_base = InMemoryKnowledgeBase("knowledge_base_data.json")
```
And internally the `InMemoryKnowledgeBase` would load the data? 
Yes, makes sense. Will change it. Would also allow the user to reload the data if the data is updated.

File: docs/core/knowledge-bases.rst (hunk 10 above)
Comment (amn41) [reply docs/core/knowledge-bases.rst]: ok. does it work to then mention that every item in the KB should have a `name` and `id` field, and if not, they should read about customizing below?

File: examples/knowledgebasebot/data/nlu.md (hunk 15 above)
Comment (tabergma) [reply examples/knowledgebasebot/data/nlu.md]: I would like to keep `object_type`. I think `type` is too generic. Would not be clear from the name that it is related to the knowledge base object.

File: docs/core/knowledge-bases.rst [hunk 25]
```diff
@@ -0,0 +1,416 @@
...
+          "9": lambda l: l[8],
+          "10": lambda l: l[9],
+          "ANY": lambda l: random.choice(list),
+          "LAST": lambda l: l[-1],
+      }
+
+  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
+  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.
+
+Defining the NLU Data
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]: Let's start this section with a brief summary:

- we are going to introduce a new intent, ``query_knowledge_base``
- we are going to annotate `mention` entities so that our model detects indirect mentions of entities like 'the first one'
- we will use synonyms extensively

File: docs/core/knowledge-bases.rst [hunk 26]
```diff
@@ -0,0 +1,416 @@
...
+- ``object_type``: Whenever the user is talking about a specific object type from your knowledge base, the type should
+  be marked as entity in our NLU data. Use :ref:`entity_synonyms` to map, for example, "restaurants" to the correct
+  object type listed in the knowledge base, e.g. "restaurant".
+- ``mention``: If the user refers to an object via "the first one", "that one", or "it", you should mark those terms
+  as ``mention``. We also use :ref:`entity_synonyms` to map some of the mentions to symbols. More on that in
+  :ref:`resolve_mentions`.
+- ``attribute``: All attribute names defined in your knowledge base should be marked as ``attribute`` in the NLU data.
+  Again, use :ref:`entity_synonyms` to map variations of an attribute name to the one used in the knowledge base.
+
+Don't forget to add those entities to your domain file (as entities and slots).
```
Comment (amn41) [inline docs/core/knowledge-bases.rst]:
```suggestion
Remember to add those entities to your domain file (as entities and slots).
```

Comment (amn41) [reply docs/core/knowledge-bases.rst]: do they need to be slots? should the be unfeaturized? let's include a snippet (possibly from the example)

File: docs/core/knowledge-bases.rst (hunk 19 above)
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: not really the KB, but more the smart usage of slots

File: docs/core/knowledge-bases.rst [hunk 27]
```diff
@@ -0,0 +1,496 @@
...
+   :local:
+
+
+Knowledge base actions enable you to handle the following kind of dialogues:
+
+.. image:: ../_static/images/knowledge-base-example.png
+
+As you can see users might want to obtain detailed information about certain objects, such as restaurants or hotels,
+during a conversation.
+They want to obtain details, such as, if an restaurant has outside seating or how expensive the restaurant is.
```
Comment (Ghostvv) [inline docs/core/knowledge-bases.rst]:
```suggestion
They want to obtain details, such as, if a restaurant has outside seating or how expensive the restaurant is.
```

File: docs/core/knowledge-bases.rst [hunk 28]
```diff
@@ -0,0 +1,496 @@
...
+``ActionQueryKnowledgeBase``.
+This is a single actions which contains the logic to query a knowledge base for objects and their attributes.
+When a restaurant is mentioned indirectly, for example using a phrase like ``the first one`` or ``that restaurant``,
+this action is able to figure out which restaurant the user is referring to.
+You can find a complete example in ``examples/knowledge_base_bot``.
+
+.. note::
+   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: I've commented later on a proposed structure, where this tutorial link would be included much later in the customization section.

File: docs/core/knowledge-bases.rst [hunk 29]
```diff
@@ -0,0 +1,496 @@
...
+
+
+.. _customize_in_memory_knowledge_base:
+
+
+Customize your InMemoryKnowledgeBase
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The class ``InMemoryKnowledgeBase`` inherits ``KnowledgeBase``.
+You can customize your ``InMemoryKnowledgeBase`` by overwriting the following functions:
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: Two ways of customizations:
1. If you don't follow all assumptions we made, you need to adapt the `InMemoryKnowledgeBase`. E.g. not all your objects have an `id` field or a `name` field.
2. If you want to use a graph database or any other database in general, you need to create a new class that inherits `KnowledgeBase`.

File: docs/core/knowledge-bases.rst [hunk 30]
```diff
@@ -0,0 +1,496 @@
...
+Knowledge Base Actions
+======================
+
+.. edit-link::
+
+.. warning::
+   This feature is experimental.
+   We introduce experimental features to get feedback on implementations, so do please try it out.
+   However, the functionality might be changed or removed in the future.
+   If you have feedback (positive or negative) please comment in the `community forum <https://forum.rasa.com>`_.
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: No, but we could create one. But maybe first release the feature, then create a forum post so that we can link to the documentation, and finally update the documentation to link to the correct forum post.

File: docs/core/knowledge-bases.rst [hunk 31]
```diff
@@ -0,0 +1,496 @@
...
+          "10": lambda l: l[9],
+          "ANY": lambda l: random.choice(list),
+          "LAST": lambda l: l[-1],
+      }
+
+  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
+  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.
+
+
+Creating your own Knowledge Base Implementation
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: I like the structure 👍 Thanks! Will update the documentation.

File: docs/core/knowledge-bases.rst (hunk 30 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: Yes, I suppose that's probably the best way to do it.

File: docs/core/knowledge-bases.rst [hunk 32]
```diff
@@ -0,0 +1,566 @@
...
+``get_attributes_of_object()``.
+You can also customize your knowledge base further, for example, by adapting the methods mentioned in the section
+:ref:`customize_in_memory_knowledge_base`.
+
+.. note::
+   We wrote a `blog post <https://blog.rasa.com/set-up-a-knowledge-base-to-encode-domain-knowledge-for-rasa/>`_
+   that explains how you can set up your own knowledge base.
+
+
+Creating your own Knowledge Base Actions
```
Comment (erohmensing) [inline docs/core/knowledge-bases.rst]: Should probably go above creating your own knowledge base -- feel like the level of customisation would make most sense as

1. Customize actionqueryknowledgebase 
2. Create your own knowledge base action 
3. Customize the in memory knowledge base 
4. Create your own knowledge base

File: docs/core/knowledge-bases.rst [hunk 33]
```diff
@@ -0,0 +1,556 @@
...
+
+- User: `What is the cuisine of PastaBar?`
+- Bot: `PastaBar has an Italian cuisine.`
+- User: `Does it have wifi?`
+- Bot: `Yes.`
+- User: `Can you give me an address?`
+
+In the second utterance of the user, the user refers to "PastaBar" by the word "it".
+If the NER detected "it" as the entity ``mention``, the knowledge base action would resolve it to the last mentioned
+object in the conversation, e.g. "PastaBar".
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: What do you mean? The NER should detect "it" as `mention`. Otherwise the action might not be able to resolve it to the last mentioned object (in case `use_last_object_mention` is set to False).

File: docs/core/knowledge-bases.rst [hunk 34]
```diff
@@ -0,0 +1,566 @@
...
+
+Customize your `ActionQueryKnowledgeBase`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can overwrite the following two functions of `ActionQueryKnowledgeBase`:
+
+- ``utter_objects()``
+- ``utter_attribute_value()``
+
+``utter_objects()`` is used when the user requested the bot to list some objects.
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: If I want to overwrite those functions, I would need to create a new knowledge base. I don't want that. I added doc strings to all methods and the `InMemoryKnowledgeBase` is basically an example implementation of the `KnowledgeBase` class. I think that is enough.
In the knowledgebasebot example I'm overwriting one of the functions. So, it should be clear how to do it.

File: docs/core/knowledge-bases.rst [hunk 35]
```diff
@@ -0,0 +1,556 @@
...
+      - mention
+      - attribute
+
+    slots:
+      object_type:
+        type: text
+      mention:
+        type: text
+      attribute:
+        type: text
```
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: I think they should be featurized as they indicate the query knowledge base intent quite well.

[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: But it is the entities featurization that does that, not the slot featurization right? 

If we keep them as text, maybe we should add a note to remember to put the slot events in the stories (but then there would be different slots being filled depending on what the input is that cause the same action -- which leads to more stories for all of the different combos, right?

File: docs/core/knowledge-bases.rst (hunk 33 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: Ah okay. I just mean when creating the NLU data, then we should also be tagging `it` as a mention? e.g. in L160:
```
what [cuisine](attribute) is it?
```
we don't tag `it` as a mention

File: docs/core/knowledge-bases.rst (hunk 35 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: not sure about entities featurization vs slot featurization... will figure it out. Would be helpful if we could keep them as unfeaturized slots as that would simplify the stories for sure.

File: docs/core/knowledge-bases.rst (hunk 34 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: > If I want to overwrite those functions, I would need to create a new knowledge base. I don't want that. I added doc strings to all methods and the InMemoryKnowledgeBase is basically an example implementation of the KnowledgeBase class. I think that is enough.

Got it. Maybe we should link to the code where those methods are found

> In the knowledgebasebot example I'm overwriting one of the functions. So, it should be clear how to do it.

Similarly, reference that there is an example of this in the knowledgebasebot example since we're pretty far down on the page by now

File: docs/core/knowledge-bases.rst (hunk 33 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: good catch! will add it.

[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: makes sense! another couple of examples of mentions like `it` and `that one` would be helpful

File: docs/core/knowledge-bases.rst (hunk 35 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: @Ghostvv might have insight on this

[the remark this answers is not in the document]
Comment (tabergma) [reply docs/core/knowledge-bases.rst]: I guess, you are right. Entities are added as features to the intent classification anyway. So, nothing needs to be done here. If slot are featurized, they are used as features for predicting the next action. That is actually not needed. So, we can change the types to unfeaturized. (Finally, I understood it 😃 )

File: docs/core/knowledge-bases.rst [hunk 36]
```diff
@@ -0,0 +1,581 @@
...
+attribute of an object.
+The ``ActionQueryKnowledgeBase`` can handle both of those requests.
+Other requests, such as comparison between objects, are currently not supported.
+
+Another thing you may have noticed is, that we marked different kind of entities in the NLU data.
+If you want to use ``ActionQueryKnowledgeBase``, you need to specify the following entities:
+
+- ``object_type``: Whenever the user is talking about a specific object type from your knowledge base, the type should
+  be marked as entity in our NLU data. Use :ref:`entity_synonyms` to map, for example, "restaurants" to the correct
+  object type listed in the knowledge base, e.g. "restaurant".
```
Comment (erohmensing) [inline docs/core/knowledge-bases.rst]: Do they also need to add the 
```
##synonym: 1
- the first one

##synonym: outside-seating
- outside seating
```
etc. to the NLU data?

Comment (tabergma) [reply docs/core/knowledge-bases.rst]: You can, but you don't need to. The bot is working just fine without defining it. As explained in the `entity_synonyms` section, two ways exists to express synonyms. I guess we can leave it as it is.

Comment (erohmensing) [reply docs/core/knowledge-bases.rst]: Got it. Just didn't know exactly how synonyms work 😃
